[V1][CUDA Graph] Fix attention metadata tensor sizes for padded batches

ayushsatyam146 opened the pull request:
Fixes #23789. Move CUDA graph padding before attention metadata construction to ensure tensor sizes match execution environment. Fixes crashes in FlashMLA and other attention backends caused by metadata/execution size mismatches.

- Pass padded token count to _prepare_inputs()
- Update slot_mapping and num_actual_tokens to use padded sizes
- Eliminates need for backend-specific workarounds

LucasWilkinson commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -838,12 +842,11 @@ def _prepare_inputs(
                 # Fill unused with -1. Needed for reshape_and_cache in full cuda
                 # graph mode.
-                blk_table.slot_mapping[total_num_scheduled_tokens:].fill_(-1)
+                blk_table.slot_mapping[num_input_tokens:].fill_(-1)
```
I think we will still want to set the padded tokens (i.e. from `total_num_scheduled_tokens` to `num_input_tokens`) to -1

LucasWilkinson reviewed (commented):
Thanks for taking this on! I think this might actually be a bit trickier then I initially thought; couple of comments

1) if we are dispatching to full cudagraphs; then we should pad `num_reqs` to match what the cudagraph was captured for since this information is used by the attention backends too; this may require changes to the `CudagraphDispatcher`
2) if we are dispatching to a piecewise CUDA-graph we don't need to pad `num_reqs` and its probably better not to (to avoid wasted work in attention)

LucasWilkinson commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -679,6 +687,12 @@ def _prepare_inputs(
         max_num_scheduled_tokens = max(tokens)
 
+        if num_reqs > actual_num_reqs:
+            padded_tokens = np.zeros(num_reqs - actual_num_reqs,
```
it might be faster to do something like
```
if num_reqs > actual_num_reqs:
      num_scheduled_tokens = np.zeros(num_reqs, dtype=np.int32)
      num_scheduled_tokens[:actual_num_reqs] = tokens # you may need to covert to an array? not sure
else:
      num_scheduled_tokens = np.array(tokens, dtype=np.int32)
max_num_scheduled_tokens = max(num_scheduled_tokens)
```

LucasWilkinson commented on vllm/v1/cudagraph_dispatcher.py:
```diff
@@ -103,19 +130,20 @@ def dispatch(
-        if non_uniform_key in self.cudagraph_keys[CUDAGraphMode.PIECEWISE]:
-            return CUDAGraphMode.PIECEWISE, non_uniform_key
+        # Check for full cudagraphs.
+        for key in self.cudagraph_keys[CUDAGraphMode.FULL]:
```
since this is on the hot-path I think we should try to keep these lookup O1

I think we can assume that for a given number of tokens theres only request count so we could make `self.cudagraph_keys` be of type 
```
CudaGraphKey: TypeAlias = tuple[int, bool] # (num_tokens, uniform_decode)
cudagraph_keys: dict[CUDAGraphMode, dict[CudaGraphKey, BatchDescriptor]]
```

then I think we could something like
```
if (num_tokens, uniform_decode) in self.cudagraph_keys[CUDAGraphMode.FULL]:
       padded_batch_descriptor = self.cudagraph_keys[CUDAGraphMode.FULL][(num_tokens, uniform_decode))
       if padded_batch_descriptor.num_reqs >= num_reqs:
             return CUDAGraphMode.FULL, padded_batch_descriptor

if (num_tokens, False) in self.cudagraph_keys[CUDAGraphMode.PIECEWISE]:
       padded_batch_descriptor = self.cudagraph_keys[CUDAGraphMode.PIECEWISE][(num_tokens, False))
       if padded_batch_descriptor.num_reqs >= num_reqs:
             return CUDAGraphMode.PIECEWISE, padded_batch_descriptor
```
Theres probably a better way of writing this just trying to demonstrate the idea

ayushsatyam146 commented:
Hi @LucasWilkinson, I addressed your comments, PTAL thanks!

fhl2000 commented:
<img width="1401" height="631" alt="image" src="https://github.com/user-attachments/assets/57e8dfb9-2c5d-40b6-9ff7-eca170b05c6b" />
 From the CI, it seems that there is an inconsistency in the input of varlen attention ops.  It is on the eager part of a piecewise cudagraph, so you can easily check which input has the wrong shape for padded size 112 (22937 is of [112, 8, 2, 128]) instead of the actual bs 111. I think in piecewise cudagraph, attn_metadata doesn't need to be padded based on previous behaviour.

fhl2000 reviewed (commented):
Hi @ayushsatyam146, the implementation is generally good. My only concern is whether we really need `num_reqs` as part of the cudagraph_dispatching process, as it looks confusing to me.  Leave some of my thoughts below.

ayushsatyam146 commented on vllm/forward_context.py:
```diff
@@ -34,6 +34,7 @@ class BatchDescriptor(NamedTuple):
     """
     num_tokens: int
     uniform_decode: bool = False
+    num_reqs: Optional[int] = None
     """
     False can also be used for an uniform decode batch to dispatch to the 
     cudagraph supporting non-uniform batches.
```
Should I wait for this PR to get merged in that case @fhl2000?

ayushsatyam146 replied:
cc: @LucasWilkinson thoughts?

fhl2000 replied:
> Should I wait for this PR to get merged in that case. 

No nessasarliy. In that case, I suggest the padded num_reqs can be directly computed or queried from somewhere (of course, it could also be a helper function of the cudagraph_dispatcher since the padding is cudagraph related), rather than making it an item of BatchDescriptor.

LucasWilkinson replied:
I think it's fines to have a `CUDAGraphKey` and a `BatchDescriptor` combo; i.e. a `CUDAGraphKey` that is a subset of `BatchDescriptor` and used for lookup.

>  I suggest the padded num_reqs can be directly computed 

I think this a bit hard to read making it more error prone; I think its just better to explicitly record the `num_reqs` for the batch when it was recorded since it makes padding to num_reqs and checking support more readable and explicit.

> or queried from somewhere (of course, it could also be a helper function of the cudagraph_dispatcher since the padding is cudagraph related)

This is pretty much exactly what is happening now; querying the requests via the `BatchDescriptor`. But by saving the whole object it just makes the return cleaner and makes it more extensible if we need to have other properties in the future that may not live in the key.

LucasWilkinson commented on vllm/v1/cudagraph_dispatcher.py:
```diff
@@ -84,11 +90,43 @@ def initialize_cudagraph_keys(self, cudagraph_mode: CUDAGraphMode,
+                                    num_reqs=num_reqs))
+
+        if cudagraph_mode.mixed_mode() == CUDAGraphMode.FULL:
+            for bs in self.compilation_config.cudagraph_capture_sizes:
+                num_reqs = self._calculate_num_reqs_for_tokens(
+                    bs, uniform_decode_query_len, False)
                 self.add_cudagraph_key(
                     CUDAGraphMode.FULL,
-                    BatchDescriptor(num_tokens=bs, uniform_decode=True))
+                    BatchDescriptor(num_tokens=bs,
+                                    uniform_decode=False,
+                                    num_reqs=num_reqs))
```
agreed ➕

LucasWilkinson commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1764,10 +1777,41 @@ def _preprocess(
+        uniform_decode = (max_query_len == self.uniform_decode_query_len) and (
+            num_scheduled_tokens == self.input_batch.num_reqs * max_query_len)
+
+        initial_batch_descriptor = BatchDescriptor(
+            num_tokens=num_input_tokens,
+            uniform_decode=uniform_decode,
+            num_reqs=self.input_batch.num_reqs)
+
+        cudagraph_runtime_mode, batch_descriptor = \
+            self.cudagraph_dispatcher.dispatch(initial_batch_descriptor)
+
+        final_num_reqs = self.input_batch.num_reqs
+        if (cudagraph_runtime_mode == CUDAGraphMode.FULL
+                and batch_descriptor is not None
+                and batch_descriptor.num_reqs is not None):
+            final_num_reqs = batch_descriptor.num_reqs
+            num_input_tokens = batch_descriptor.num_tokens
+        elif cudagraph_runtime_mode == CUDAGraphMode.PIECEWISE:
+            num_input_tokens = unpadded_num_input_tokens
+
+        # Prepare the decoder inputs.
+        (attn_metadata, logits_indices, spec_decode_metadata,
+         num_scheduled_tokens_np, spec_decode_common_attn_metadata,
+         _) = self._prepare_inputs(scheduler_output, num_input_tokens,
+                                   final_num_reqs)
```
agreed; this seems strange

ayushsatyam146 committed:
[V1][CUDA Graph] Fix attention metadata tensor sizes for padded batches

Currently, cudagraph padding is applied after attention metadata is created,
causing dimension mismatches between metadata and actual tensors. This leads
to issues in attention backends like FlashMLA and FlashAttn.

This change moves the padding calculation before metadata construction and
adds support for request-level padding in full cudagraph mode:

- Add num_reqs field to BatchDescriptor for request dimension tracking
- Calculate padding requirements before _prepare_inputs() call
- Support both token and request padding for full cudagraphs
- Only pad tokens for piecewise cudagraphs to avoid wasted computation
- Fix slot mapping padding to only fill padded region

Signed-off-by: ayushsatyam146 <ayushsatyam146@gmail.com>

LucasWilkinson committed:
review comments

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
Merge branch 'main' into cudagraph-fix

LucasWilkinson committed:
format

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
Merge branch 'main' into cudagraph-fix

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
cleanup

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
cleanup

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
more cleanup

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
Merge remote-tracking branch 'origin/main' into cudagraph-fix

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
cleanup

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
clean up

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
no need for lora change

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

ProExpertProg commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2046,18 +2024,10 @@ def _pool(
-            return self.vllm_config.pad_for_cudagraph(num_scheduled_tokens)
-
-        # Eager mode.
+        """
```
I think sequence parallelism tests are not running in CI, could you move them or wait for #26038 so we can make sure that path still works?

ProExpertProg commented on vllm/v1/cudagraph_dispatcher.py:
```diff
@@ -113,21 +162,19 @@ def dispatch(
-            # check if key exists for full cudagraph
-            if batch_descriptor in self.cudagraph_keys[CUDAGraphMode.FULL]:
-                return CUDAGraphMode.FULL, batch_descriptor
+        num_tokens, uniform_decode = (
+            batch_descriptor.num_tokens,
+            batch_descriptor.uniform_decode,
+        )
```
Can we make this a new data structure? Or can we just use `BatchDescriptor` with `num_reqs=None`?

ProExpertProg commented on vllm/v1/cudagraph_dispatcher.py:
```diff
@@ -113,21 +162,19 @@ def dispatch(
-        # piecewise cudagraph
-        if non_uniform_key in self.cudagraph_keys[CUDAGraphMode.PIECEWISE]:
-            return CUDAGraphMode.PIECEWISE, non_uniform_key
+        for mode, key in candidates:
```
I think this makes the logic a bit more confusing so we should add comments and add to the dispatching docs. What we're saying here is that if we have a uniform cudagraph that cannot handle the required number of requests, we downgrade to piecewise? Will that happen or do we make sure it doesn't ahead of time? Just worried about this kind of "hidden fallback"

ProExpertProg commented on vllm/v1/cudagraph_dispatcher.py:
```diff
@@ -94,12 +117,38 @@ def initialize_cudagraph_keys(
+
         self.keys_initialized = True
 
+    def calculate_num_reqs_for_tokens(
```
I don't think I understand what this does. It gives the maximum possible number of non-empty requests? But I thought we also have requests of length 0 sometimes?

ProExpertProg commented on vllm/v1/cudagraph_dispatcher.py:
```diff
@@ -1,9 +1,16 @@
+from vllm.logger import init_logger
+
+logger = init_logger(__name__)
+CUDAGraphKey: TypeAlias = tuple[int, bool]
```
Can we make this a dataclass with utilities for easier and more understandable handling as well as more docs?

ProExpertProg commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -3385,50 +3454,26 @@ def _dummy_run(
 
             if ubatch_slices is not None:
                 # Adjust values to reflect a single ubatch.
                 # TODO(sage,lucas): this is cruft that should be addressed in
```
Is this the padding refactor? OR is that a different PR?

LucasWilkinson replied:
that'll likely be a different PR

ProExpertProg commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -3260,20 +3298,52 @@ def _dummy_run(
+        max_num_scheduled_tokens = int(num_scheduled_tokens.max())
+
+        # Step 1: Pad for sequence parallelism (TP-related padding)
+        padded_num_tokens = self._get_num_input_tokens(total_num_scheduled_tokens)
```
Oh I see - can we call this method `pad_for_sequence_parallelism` then?

ProExpertProg commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -3260,20 +3298,52 @@ def _dummy_run(
-        num_tokens_after_padding = num_tokens
         if num_tokens_across_dp is not None:
-            num_tokens_after_padding = int(num_tokens_across_dp[0])
+            padded_num_tokens = int(num_tokens_across_dp[0])
```
Is `[0]` always the current rank? Or are they all the same?

LucasWilkinson replied:
they are all the same; either way behavior is unchanged by this PR we need to clean-up the padding in a following PR

ProExpertProg reviewed (commented):
I don't think I understand the padding logic, could we add some more comments?

LucasWilkinson committed:
review comments

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

SageMoore reviewed (commented):
Hello @LucasWilkinson. Could you briefly explain how the cudagraph padding is applied with these changes. I see the padding logic deleted from `_get_num_input_tokens` and the general "flow" of the `execute_model `code seems to hint that we expect the cudagraph dispatcher to round up the number of tokens in the BatchDescriptor to the next cudagraph size, but I don't see that done anywhere.

AIUI, when `_maybe_dispatch_cudagraph` is called, `padded_num_tokens` is equal to SP padding + DP padding. I don't see anything accounting for cudagraph padding, though.

LucasWilkinson committed:
Merge remote-tracking branch 'origin/main' into cudagraph-fix

LucasWilkinson committed:
more refactoring

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson reviewed (changes requested):
This PR is getting too large; will break it apart into smaller PRs with this being the final one

LucasWilkinson committed:
unifiy build attention metadata

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
clean-up

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
refactor

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
wip

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
cleanup

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
cleanup

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
fix

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
cleanup

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
cleanup

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
fix

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
clean up

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
fix docs error

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
Merge remote-tracking branch 'nm/lwilkinson/seperate-build-attn-metadata' into cudagraph-fix

LucasWilkinson committed:
Fix merge conflicts: add missing imports and fix indentation

- Add missing Optional import to forward_context.py
- Add missing NamedTuple import to cudagraph_dispatcher.py
- Add missing Optional import to gpu_model_runner.py
- Fix indentation in gpu_model_runner.py (try-except block)

LucasWilkinson committed:
Merge remote-tracking branch 'origin/main' into pr/ayushsatyam146/24002

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
wip

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
wip

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
Merge branch 'lwilkinson/pad-before-metadata' into pr/ayushsatyam146/24002

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
update docs

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>

LucasWilkinson committed:
cleanup

Signed-off-by: Lucas Wilkinson <lwilkins@redhat.com>
